crates/wasmparser/src/readers/core/data.rs

```diff
@@ -112,32 +105,47 @@ impl<'a> DataSectionReader<'a> {
     {
         let segment_start = self.reader.original_position();
 
+        // The current handling of the flags is largely specified in the `bulk-memory` proposal,
+        // which at the time this commend is written has been merged to the main specification
+        // draft.
+        //
+        // Notably, this proposal allows multiple different encodings of the memory index 0. `00`
+        // and `02 00` are both valid ways to specify the 0-th memory. However it also makes
+        // another encoding of the 0-th memory `80 00` no longer valid.
+        //
+        // We, however maintain this by parsing `flags` as a LEB128 integer. In that case, `80 00`
+        // encoding is parsed out as `0` and is therefore assigned a `memidx` 0, even though the
+        // current specification draft does not allow for this.
+        //
+        // See also https://github.com/WebAssembly/spec/issues/1439
         let flags = self.reader.read_var_u32()?;
```

Comment:
This is actually a bug, because we end up accepting encodings like `80 01` (which parses to flag 1, which isn't allowed by the current specification draft text), but it also allows for backwards compatible behaviour for a `WebAssembly 1.0` definition of `data` section entry. See the referenced issue for more context.

Reply:
Oh this was only here because it was necessary to pass spec tests before the bulk memory proposal was merged, nowadays it's fine to change this to `read_u8` and match on the byte read to determine the kind of segment.

Reply:
There still are a number of tests which test for the `80 00` encoding and start failing if this is replaced with `read_u8`. For example https://github.com/WebAssembly/testsuite/blob/d42da0117f7a93c6a9127e2b9eec64749152c4c1/binary-leb128.wast#L24-L31. I'll leave as-is for now, but would be good to eventually have some eyes on it.

Reply:
Ah I did not realize that, yeah in that case sounds good to leave here. I'll see if I can do a patch to the spec test suite to remove those tests.

Otherwise though FWIW I believe that the breakage relative to the original specification was deliberate. In practice I don't think anyone actually used the over-long encodings for memory indices so it was seen as a reasonable opportunity to just go ahead and use a leading byte instead of having a weird encoding scheme.